### PR TITLE
Add rate limiter.

### PR DESCRIPTION
**Pull request by jnb**
Allow more control over HAProxy restart rate.

### CODE REVIEW

File: lib/synapse/rate_limiter.rb [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  class RateLimiter
+    include Logging
+
+    def initialize(path)
+      @path = path
+
+      # Our virtual clock
+      @time = 0
+
+      # Average at most one restart per minute, with the ability to burst to
```
Comment (jnb) [inline lib/synapse/rate_limiter.rb:13]: We can tune these params.

File: lib/synapse/rate_limiter.rb [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        @tokens = [@tokens + 1, @max_tokens].min
+      end
+
+      # Save state out to disk
+      data = {
+        "time" => @time,
+        "last_restart_time" => @last_restart_time,
+        "tokens" => @tokens,
+      }
+      begin
```
Comment (jnb) [inline lib/synapse/rate_limiter.rb]: I should check for @path.nil? here.

File: lib/synapse/rate_limiter.rb [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    def initialize(path)
+      @path = path
+
+      # Our virtual clock
+      @time = 0
+
+      # Average at most one restart per minute, with the ability to burst to
+      # two restarts per minute
+      @tokens = 0
+      @max_tokens = 2
```
Comment (jolynch) [inline lib/synapse/rate_limiter.rb:16]: It seems like these should be options? Perhaps for the next iteration, but I could see it being a problem if we effectively only restart once every minute.

File: lib/synapse.rb [hunk 4]
```diff
@@ -33,6 +34,8 @@ def initialize(opts={})
       # configuration is initially enabled to configure on first loop
       @config_updated = true
 
+      @rate_limiter = RateLimiter.new(opts['rate_limiter_path'])
```
Comment (jolynch) [inline lib/synapse.rb:37]: Think we should log something if this path doesn't exist? If I understand correctly that would mean that restarting synapse would restart haproxy every time?

**Comment by jolynch** [comment]
:+1: 
Mostly minor comments, feel free to follow up in the next iteration.

I love this token bucket stuff

File: lib/synapse/rate_limiter.rb [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        rescue
+        end
+      end
+    end
+
+    def proceed?
+      # Is there an available token?
+      if @tokens > 0
+        # Has it been sufficiently long since our last restart?
+        if @time - @min_restart_period >= @last_restart_time
```
Comment (jnb) [inline lib/synapse/rate_limiter.rb:65]: I'd like to try adding some jitter here.

File: lib/synapse.rb (hunk 4 above)
Comment (jnb) [inline lib/synapse.rb:37]: We currently log from initialize() if the state file can't be read.  I think this is sufficient?

**Comment by jolynch** [comment]
:+1:

**Pull request merged**